dpcpp/solver/batch_bicgstab_kernels.hpp.inc

```diff
@@ -0,0 +1,455 @@
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+template <const bool sg_kernel_all, typename BatchMatrixType_entry,
+          typename ValueType>
+__dpct_inline__ void initialize(
+    const int num_rows, const BatchMatrixType_entry& mat_global_entry,
+    const ValueType* const b_global_entry,
+    const ValueType* const x_global_entry, ValueType& rho_old, ValueType& omega,
+    ValueType& alpha, ValueType* const x_shared_entry,
+    ValueType* const r_shared_entry, ValueType* const r_hat_shared_entry,
+    ValueType* const p_shared_entry, ValueType* const v_shared_entry,
+    ValueType* const p_hat_shared_entry,
+    typename gko::remove_complex<ValueType>& rhs_norm,
+    typename gko::remove_complex<ValueType>& res_norm,
+    sycl::nd_item<3> item_ct1)
+{
+    auto sg = item_ct1.get_sub_group();
+    const auto sg_id = sg.get_group_id();
+    const auto tid = item_ct1.get_local_linear_id();
+    const auto group_size = item_ct1.get_local_range().size();
+    const auto group = item_ct1.get_group();
+
+    rho_old = one<ValueType>();
+    omega = one<ValueType>();
+    alpha = one<ValueType>();
+
+    // copy x from global to shared memory
+    // r = b
+    for (int iz = tid; iz < num_rows; iz += group_size) {
+        x_shared_entry[iz] = x_global_entry[iz];
+        r_shared_entry[iz] = b_global_entry[iz];
+    }
+    item_ct1.barrier(sycl::access::fence_space::local_space);
+
+    // r = b - A*x
+    advanced_apply_kernel(static_cast<ValueType>(-1.0), mat_global_entry,
+                          x_shared_entry, static_cast<ValueType>(1.0),
+                          r_shared_entry, item_ct1);
+    item_ct1.barrier(sycl::access::fence_space::local_space);
```

Comment:
you need to ensure the x, r are indeed in the local shared memory, or it may get the old result from global memory.

Reply:
I dont think that is necessary. If it is in global memory, all the subgroup reads and writes should be consistent. A variable does not switch between local shared and global. If it is global memory, it always remains in global and if it is in local shared it always remains in local shared until the end when it is copied back. Only when it is in local shared memory, consistency between different work-items needs to be enforced with a `fence_space::local_space`

Reply:
you are probably looking for a `group_barrier(item_ct1)`? As far as I understand the documentation on the [Codeplay website](https://developer.codeplay.com/products/computecpp/ce/2.11.0/api-reference/namespacecl_1_1sycl_1_1access), this applies only to local memory.

Reply:
Is there even a `group_barrier` function ? I think `local_space` is the local shared memory, so enforcing a fence on the local memory should already ensure that all the work-items in the workgroup are synchronized ?

Reply:
According to the SYCL 2020 standard there is. There is a difference beetween visibility of memory operations between different threads and their execution synchronization. The SYCL standard doesn't say anything about global memory coherence when synchronizing local memory, so we can't rely on it. The group_barrier function has a stronger contract. It may work on specific hardware, but may still give unexpected results on other hardware.

Reply:
In practical terms, subgroups (and likely also workgroups) will use the same L1 cache, so they will likely have the same view of global memory, but the compiler may still be allowed to reorder operations that it knows happen in global memory and not local memory.

Reply:
Indeed, it seems that with SYCL 2020, `group_barrier()` is the equivalent of `__syncthreads()`, or the equivalent `fence_space` seems to be `fence_space::global_and_local`. But it [does seem that both `group_barrier()` and `fence_space::global_and_local` are significantly slower than `fence_space::local_space`](https://github.com/intel/llvm/issues/7069). IIRC, Phuong also observed this in her experiments.

Reply:
But I think for correctness, I will change this to `fence_space::global_and_local`, and we can investigate the performance at a later point.

Reply:
synchronize with global and local should definitely be slower than only with local without doubt.
If we do not adapt it with the shared/global switch, we can only use global_and_local for correctness unfortunately.
using local will work if all threads always access the same global address in whole kernel.
I think it is usually true except for norm/dot.